sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.redis;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.ArrayList;
+
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import redis.embedded.RedisServer;
+
+/**
+ * Test on the Redis IO.
+ */
+public class RedisIOTest {
+
+  @Rule public TestPipeline pipeline = TestPipeline.create();
+
+  private EmbeddedRedis embeddedRedis;
+
+  @Before
+  public void before() throws Exception {
+    embeddedRedis = new EmbeddedRedis();
+  }
+
+  @After
+  public void after() throws Exception {
+    embeddedRedis.close();
+  }
+
+  @Test
+  public void testWriteRead() throws Exception {
+    RedisConnectionConfiguration connection = createConnection(embeddedRedis);
+
+    ArrayList<KV<String, String>> data = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      KV<String, String> kv = KV.of("key " + i, "value " + i);
+      data.add(kv);
+    }
+    PCollection<KV<String, String>> write = pipeline.apply(Create.of(data));
+    write.apply(RedisIO.write().withConnectionConfiguration(connection));
+
+    pipeline.run();
+
+    PCollection<KV<String, String>> read = pipeline.apply("Read",
```

Comment:
You need a new pipeline after you've run the current one - using a pipeline after it's been run is undefined behavior. I'm surprised we don't enforce this with runtime errors...